code/obj/machinery/sleeper.dm

```diff
@@ -128,146 +127,97 @@
-
-			else
-				dat += "<HR>The sleeper is unoccupied."
-
-		else
-			dat += {"<font color='red'><b>ERROR:</b> No sleeper detected!</font><br>
-					<br><A href='?src=\ref[src];refresh=1'>Refresh Connection</A>"}
-
-		user.Browse(dat, "window=sleeper")
-		onclose(user, "sleeper")
-
-		return
-
-	Topic(href, href_list)
-		/*if (..())
-			return*/
-		if (!isturf(src.loc) && !istype(src.loc, /obj/machinery/sleeper))
-			return
-		if ((src.our_sleeper && src.our_sleeper.occupant == usr) || usr.getStatusDuration("stunned") || usr.getStatusDuration("weakened") || usr.stat || usr.restrained())
-			return
-		if (!issilicon(usr) && !in_range(src, usr))
+	ui_status(mob/user)
+		var/use_obj = src
+		if (src.loc == src.our_sleeper) // port-a-medbay
+			use_obj = src.loc
+			if (user in use_obj)
+				return UI_CLOSE
+		return min(
+			tgui_broken_state.can_use_topic(use_obj, user),
+			tgui_physical_state.can_use_topic(use_obj, user),
+			tgui_not_incapacitated_state.can_use_topic(use_obj, user)
+		)
+
+	ui_act(action, list/params, datum/tgui/ui, datum/ui_state/state)
+		. = ..()
+		if (.)
 			return
-
-		src.add_fingerprint(usr)
-		src.add_dialog(usr)
-
-		if (href_list["time"])
-			if (src.our_sleeper && src.our_sleeper.occupant)
-				if (isdead(src.our_sleeper.occupant))
-					usr.show_text("The occupant is dead.", "red")
-				else
-					src.timing = text2num(href_list["time"])
+		switch(action)
+			if("timer")
+				if (src.our_sleeper?.occupant && !isdead(src.our_sleeper.occupant))
+					src.timing = !src.timing
 					src.visible_message("<span class='notice'>[usr] [src.timing ? "sets" : "stops"] the [src]'s occupant alarm clock.</span>")
 					if (src.timing)
-						src.time_started = world.timeofday//realtime
+						src.time_started = TIME
 						// People do use sleepers for grief from time to time.
 						logTheThing("station", usr, src.our_sleeper.occupant, "initiates a sleeper's timer ([src.our_sleeper.emagged ? "<b>EMAGGED</b>, " : ""][src.time/10] seconds), forcing [constructTarget(src.our_sleeper.occupant,"station")] asleep at [log_loc(src.our_sleeper)].")
 					else
+						src.time = clamp(src.time + src.time_started - TIME, 0, maximum_time)
 						src.time_started = 0
 						src.wake_occupant()
+				. = TRUE
+			if("time_add")
+				if (src.our_sleeper && src.time <= maximum_time)
+					var/t = params["tp"]
+					if (t > 0 && src.timing && src.our_sleeper.occupant)
+						// People do use sleepers for grief from time to time.
+						logTheThing("station", usr, src.our_sleeper.occupant, "increases a sleeper's timer ([src.our_sleeper.emagged ? "<b>EMAGGED</b>, " : ""]occupied by [constructTarget(src.our_sleeper.occupant,"station")]) by [t] seconds at [log_loc(src.our_sleeper)].")
+					src.time = clamp(src.time + (t*10), 0, maximum_time)
+				. = TRUE
+			if("inject")
+				var/is_recharging = src.our_sleeper.no_med_spam && world.time < src.our_sleeper.no_med_spam + injection_delay
```

Comment:
Will want to add a check that `src.our_sleeper` exists prior to accessing properties of it. `?.` might work, haven't thought through the logic, so feel free to do that.